src/frontends/lean/structure_cmd.cpp

```diff
@@ -680,56 +864,71 @@ struct structure_cmd_fn {
         }
     }
 
-    expr mk_field_binder(buffer<field_decl> const & decls, expr const & type,
-                         bool typed_defaults_only = false) {
-        expr r = type;
-        unsigned i = decls.size();
-        while (i > 0) {
-            --i;
-            field_decl const & decl = decls[i];
-            if (decl.default_val && (typed_defaults_only == decl.explicit_type)) {
-                expr type  = mlocal_type(decl.local);
-                expr value = *decl.default_val;
-                if (decl.from_parent) {
-                    type  = mk_as_is(type);
-                }
-                if (typed_defaults_only)
-                    r = mk_let(mk_fresh_name(), type, value, r);
-                else
-                    r = mk_let(local_pp_name(decl.local), type, value, abstract_local(r, decl.local));
-            } else if (!typed_defaults_only) {
-                if (decl.from_parent) {
-                    r = Pi_as_is(decl.local, r);
-                } else {
-                    r = Pi(decl.local, r);
-                }
-            }
+    /* Elaborate `Pi n : T, tmp` for each field, or `let n : T := d in tmp` for each defaulted field that will not be
+     * handled by `elaborate_new_typed_default` */
+    expr elaborate_field(field_decl & decl, expr const & tmp, tele_elab elab) {
+        expr new_tmp;
+        expr type  = decl.get_type();
+        auto const & value = decl.m_default_val;
+        if (value && (!decl.m_has_new_default || is_placeholder(decl.get_type()))) {
+            new_tmp = elab(mk_let(decl.get_name(), type, *value, abstract_local(tmp, decl.m_local)));
+            decl.m_local = mk_local(decl.get_name(), let_type(new_tmp), {});
+            decl.m_default_val = let_value(new_tmp);
+            new_tmp = instantiate(let_body(new_tmp), m_subobjects ? let_value(new_tmp) : decl.m_local);
+        } else {
+            new_tmp = elab(Pi(decl.get_name(), type, abstract_local(tmp, decl.m_local), local_info(decl.m_local)));
+            decl.m_local = mk_local(decl.get_name(), binding_name(new_tmp), binding_domain(new_tmp),
+                                    binding_info(new_tmp));
+            new_tmp = instantiate(binding_body(new_tmp), decl.m_local);
         }
-        return r;
+        return new_tmp;
+    }
+
+    /* let-elaborate new default values of fields with explicit types so that they can refer to any field, even cyclically.
+     *
+     *   class Eq (α : Type) :=
+     *   (eq : α → α → Prop := λ a b, ¬ne a b)
+     *   (ne : α → α → Prop := λ a b, ¬eq a b)
+     *
+     * will be elaborated as
+     *
+     *   Π (α : Type) (eq : α → α → Prop) (ne : α → α → Prop),
```

Comment:
It isn't clear to me how this works, not essential to the PR but could be good to clarify for future readers of the code.